### PR TITLE
*: improve performance of DecodeBytes in DecodeOneToChunk

### PR DESCRIPTION
**Pull request by XuHuaiyu**
Before this commit, it will make a new byte slice in `decodeBytes` every time.
This is not worth for DecodeOneToChunk.

This commit reuses a buffered byte slice to improve
the performance of DecodeBytes in DecodeOneToChunk.

This can reduce almost 34.2%  CPU time and almost 37.2% memory usage.

Benchmark result:
BenchmarkDecodeOneToChunk-old       20000000                91.5 ns/op           43 B/op          1 allocs/op
BenchmarkDecodeOneToChunk-new     20000000                60.2 ns/op           27 B/op          0 allocs/op

PTAL @zz-jason @coocood

### CODE REVIEW

File: distsql/distsql.go [hunk 1]
```diff
@@ -205,9 +205,10 @@ func (r *selectResult) readRowsData(chk *chunk.Chunk) (err error) {
 	rowsData := r.selectResp.Chunks[r.respChkIdx].RowsData
 	maxChunkSize := r.ctx.GetSessionVars().MaxChunkSize
 	timeZone := r.ctx.GetSessionVars().GetTimeZone()
+	var bufferedBytes []byte
```
Comment (ngaut) [inline distsql/distsql.go]: s/bufferedBytes/buf/

Comment (winkyao) [reply distsql/distsql.go]: Why not save the buf in the selectResult?

File: util/codec/bytes.go [hunk 2]
```diff
@@ -116,8 +119,8 @@ func decodeBytes(b []byte, reverse bool) ([]byte, []byte, error) {
 
 // DecodeBytes decodes bytes which is encoded by EncodeBytes before,
 // returns the leftover bytes and decoded value if no error.
-func DecodeBytes(b []byte) ([]byte, []byte, error) {
-	return decodeBytes(b, false)
+func DecodeBytes(b []byte, data []byte) ([]byte, []byte, error) {
```
Comment (coocood) [inline util/codec/bytes.go]: I think `buf` is a better name.
And please update the comment of the newly added argument.

File: util/codec/codec.go [hunk 3]
```diff
@@ -482,9 +482,10 @@ func peekUvarint(b []byte) (int, error) {
 }
 
 // DecodeOneToChunk decodes one value to chunk and returns the remained bytes.
-func DecodeOneToChunk(b []byte, chk *chunk.Chunk, colIdx int, ft *types.FieldType, loc *time.Location) (remain []byte, err error) {
+// bufferedBytes is only used for DecodeBytes to avoid the cost of makeslice.
+func DecodeOneToChunk(b []byte, bufferedBytes []byte, chk *chunk.Chunk, colIdx int, ft *types.FieldType, loc *time.Location) (remain []byte, _ []byte, err error) {
```
Comment (coocood) [inline util/codec/codec.go]: Why use `_` instead of a name?

Comment (coocood) [inline util/codec/codec.go]: There are too many arguments and returned values.
How about we define a type `Decoder` that contains the `buffer`, `chunk`, and `loc`?

**Comment by shenli** [comment]
@XuHuaiyu Please fix the CI.

**Comment by XuHuaiyu** [comment]
PTAL @coocood @winkyao @zz-jason

File: util/codec/bytes.go [hunk 4]
```diff
@@ -131,8 +135,8 @@ func EncodeBytesDesc(b []byte, data []byte) []byte {
 
 // DecodeBytesDesc decodes bytes which is encoded by EncodeBytesDesc before,
 // returns the leftover bytes and decoded value if no error.
-func DecodeBytesDesc(b []byte) ([]byte, []byte, error) {
-	return decodeBytes(b, true)
+func DecodeBytesDesc(b []byte, data []byte) ([]byte, []byte, error) {
```
Comment (zz-jason) [inline util/codec/bytes.go]: s/data/buf ?

File: util/codec/codec.go [hunk 5]
```diff
@@ -481,11 +481,20 @@ func peekUvarint(b []byte) (int, error) {
 	return n, nil
 }
 
+// Decoder is mainly used to store the result of DecodeOneToChunk.
+type Decoder struct {
+	Chk *chunk.Chunk
+	// buf is only used for DecodeBytes to avoid the cost of makeslice.
+	Buf      []byte
```
Comment (zz-jason) [inline util/codec/codec.go]: no need to export ?

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason

**Pull request merged**